cmd/agent/app/throttling/throttler.go

```diff
@@ -0,0 +1,294 @@
+	perOperationRateLimiter *overrideMap
+	updateTime              time.Time
+	timeNow                 func() time.Time // For testing
+}
+
+func newAccount(options AccountOptions, timeNow func() time.Time) *account {
+	defaultRateLimiter := newTokenBucket(options, timeNow)
+	perOperationRateLimiter := newOverrideMap(options.MaxOperations, defaultRateLimiter)
+	return &account{
+		options:                 options,
+		updateTime:              timeNow(),
+		timeNow:                 timeNow,
+		perOperationRateLimiter: perOperationRateLimiter,
+	}
+}
+
+// Withdraw deducts credits from the account for the given operationName with an
+// upper limit of maxWithdrawal.
+func (a *account) Withdraw(operationName string, maxWithdrawal float64) float64 {
+	var rateLimiter *tokenBucket
+	if a.perOperationRateLimiter.Has(operationName) ||
+		a.perOperationRateLimiter.IsFull() {
+		rateLimiter = a.perOperationRateLimiter.Get(operationName).(*tokenBucket)
+	} else {
+		rateLimiter = newTokenBucket(a.options, a.timeNow)
+		a.perOperationRateLimiter.Set(operationName, rateLimiter)
+	}
+	credits := rateLimiter.Withdraw(maxWithdrawal)
+	return credits
+}
+
+// ThrottlerOptions provides values to be used in a Throttler object.
+type ThrottlerOptions struct {
+	// DefaultAccountOptions defines the default AccountOptions to use for all
+	// service accounts.
+	DefaultAccountOptions AccountOptions
+	// AccountOptionOverrides overrides DefaultAccountOptions for services with
+	// service names present in the map.
+	AccountOptionOverrides map[string]*AccountOptions
+
+	// ClientMaxBalance defines the maximum balance a client may maintain before
+	// further Withdraw calls return zero.
+	ClientMaxBalance float64
+
+	// TTL defines the time to await further updates before purging entry for
+	// all accounts and clients.
+	TTL time.Duration
+	// PurgeInterval defines the amount of time to wait between purge calls.
+	PurgeInterval time.Duration
+}
+
+// Throttler manages the relationship between service instance clients and
+// backend storage credits. Each client must withdraw credits in order to submit
+// trace data to the Jaeger agent. These credits are generated on behalf of the
+// service, of which there can be many instances, and thus many clients.
+type Throttler struct {
+	sync.Mutex
+	accounts  map[string]*account
+	clients   map[string]*client
+	options   ThrottlerOptions
+	timeNow   func() time.Time // For testing
+	ticker    *time.Ticker
+	done      chan bool
+	waitGroup sync.WaitGroup
+}
+
+// NewThrottler creates a new throttler and returns it to the caller.
+// options should specify the specific throttler options to use.
+func NewThrottler(options ThrottlerOptions) *Throttler {
+	t := &Throttler{
+		accounts: map[string]*account{},
+		clients:  map[string]*client{},
+		options:  options,
+		timeNow:  time.Now,
+	}
+	t.ticker = time.NewTicker(t.options.PurgeInterval)
+	t.done = make(chan bool)
+	t.waitGroup.Add(1)
+	go func() {
+		defer t.waitGroup.Done()
```

Comment:
ticker could be started and killed within the go routine.

Reply:
Not even going to bother killing it, just going to make it a local variable.